File: admin-panel/lib/services/lead_service.dart
import 'api_client.dart';

class Lead {
  Lead({
    required this.id,
    required this.name,
    required this.company,
    required this.email,
    required this.whatsapp,
    required this.objective,
    this.segment,
    this.monthlyVolume,
    this.teamSize,
    this.currentTools,
    this.bestContactTime,
    this.source = 'landing',
    this.status = 'new',
    this.notes,
    this.assignedTo,
    this.metadata,
    this.createdAt,
    this.updatedAt,
  });

  final String id;
  final String name;
  final String company;
  final String? segment;
  final String email;
  final String whatsapp;
  final String objective;
  final String? monthlyVolume;
  final String? teamSize;
  final String? currentTools;
  final String? bestContactTime;
  final String source;
  final String status;
  final String? notes;
  final String? assignedTo;
  final Map<String, dynamic>? metadata;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  factory Lead.fromJson(Map<String, dynamic> json) {
    DateTime? parse(dynamic value) {
      if (value == null) return null;
      if (value is DateTime) return value;
      return DateTime.tryParse(value.toString());
    }

    return Lead(
      id: (json['id'] ?? '').toString(),
      name: (json['name'] ?? '').toString(),
      company: (json['company'] ?? '').toString(),
      segment: json['segment'] as String?,
      email: (json['email'] ?? '').toString(),
      whatsapp: (json['whatsapp'] ?? '').toString(),
      objective: (json['objective'] ?? '').toString(),
      monthlyVolume: json['monthly_volume'] as String?,
      teamSize: json['team_size'] as String?,
      currentTools: json['current_tools'] as String?,
      bestContactTime: json['best_contact_time'] as String?,
      source: (json['source'] ?? 'landing').toString(),
      status: (json['status'] ?? 'new').toString(),
      notes: json['notes'] as String?,
      assignedTo: json['assigned_to']?.toString(),
      metadata: (json['metadata'] is Map<String, dynamic>)
          ? json['metadata'] as Map<String, dynamic>
          : null,
      createdAt: parse(json['created_at']),
      updatedAt: parse(json['updated_at']),
    );
  }
}

class LeadSummary {
  LeadSummary({required this.total, required this.byStatus});

  final int total;
  final Map<String, int> byStatus;

  factory LeadSummary.fromJson(Map<String, dynamic> json) {
    final raw = json['by_status'];
    final map = <String, int>{};
    if (raw is Map) {
      raw.forEach((key, value) {
        if (key == null) return;
        map[key.toString()] = (value is num) ? value.toInt() : int.tryParse('$value') ?? 0;
      });
    }
    return LeadSummary(
      total: (json['total'] is num) ? (json['total'] as num).toInt() : 0,
      byStatus: map,
    );
  }
}

class LeadService {
  LeadService({ApiClient? client}) : _client = client ?? apiClient;

  final ApiClient _client;

  Future<Lead> submitPublicLead({
    required String name,
    required String company,
    String? segment,
    required String email,
    required String whatsapp,
    required String objective,
    String? monthlyVolume,
    String? teamSize,
    String? currentTools,
    String? bestContactTime,
    String source = 'landing',
    Map<String, dynamic>? metadata,
  }) async {
    final payload = <String, dynamic>{
      'name': name.trim(),
      'company': company.trim(),
      'segment': (segment ?? '').trim().isEmpty ? null : segment!.trim(),
      'email': email.trim().toLowerCase(),
      'whatsapp': whatsapp.trim(),
      'objective': objective.trim(),
      'monthly_volume': (monthlyVolume ?? '').trim().isEmpty ? null : monthlyVolume!.trim(),
      'team_size': (teamSize ?? '').trim().isEmpty ? null : teamSize!.trim(),
      'current_tools': (currentTools ?? '').trim().isEmpty ? null : currentTools!.trim(),
      'best_contact_time':
          (bestContactTime ?? '').trim().isEmpty ? null : bestContactTime!.trim(),
      'source': source,
      if (metadata != null && metadata.isNotEmpty) 'metadata': metadata,
    };
    final result = await _client.post('/api/v1/public/leads', body: payload);
    return Lead.fromJson(Map<String, dynamic>.from(result as Map));
  }

  Future<List<Lead>> listLeads({
    String? status,
    String? search,
    int limit = 100,
    int offset = 0,
  }) async {
    final query = <String>[];
    if (status != null && status.isNotEmpty) query.add('status=${Uri.encodeQueryComponent(status)}');
    if (search != null && search.isNotEmpty) query.add('search=${Uri.encodeQueryComponent(search)}');
    query.add('limit=$limit');
    query.add('offset=$offset');
    final result = await _client.get('/api/v1/admin/leads?${query.join('&')}');
    if (result is! List) return const <Lead>[];
    return result
        .whereType<Map>()
        .map((item) => Lead.fromJson(Map<String, dynamic>.from(item)))
        .toList(growable: false);
  }

  Future<LeadSummary> summary() async {
    final result = await _client.get('/api/v1/admin/leads/summary');
    if (result is! Map) return LeadSummary(total: 0, byStatus: const {});
    return LeadSummary.fromJson(Map<String, dynamic>.from(result));
  }

  Future<Lead> updateLead({
    required String leadId,
    String? status,
    String? notes,
    int? assignedTo,
  }) async {
    final payload = <String, dynamic>{};
    if (status != null) payload['status'] = status;
    if (notes != null) payload['notes'] = notes;
    if (assignedTo != null) payload['assigned_to'] = assignedTo;
    final result = await _client.patch('/api/v1/admin/leads/$leadId', body: payload);
    return Lead.fromJson(Map<String, dynamic>.from(result as Map));
  }
}

final leadService = LeadService();

File: backend-api/app/services/lead_service.py
"""Service layer for landing-page lead capture."""
from __future__ import annotations

import json
import logging
import re
from datetime import datetime
from typing import Any

from app.config.settings import Settings
from app.domain.lead import Lead
from app.repositories.lead_repository import LeadRepository
from app.services.email_service import EmailService

logger = logging.getLogger(__name__)

_EMAIL_RE = re.compile(r"^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$")
_DIGITS_RE = re.compile(r"\D+")
_ALLOWED_STATUSES = {"new", "contacted", "qualified", "won", "lost"}


def _normalize_whatsapp(value: str) -> str:
    digits = _DIGITS_RE.sub("", value or "")
    if not digits:
        return ""
    if len(digits) <= 11 and not digits.startswith("55"):
        digits = "55" + digits
    return digits


def _normalize_email(value: str) -> str:
    return (value or "").strip().lower()


def _row_to_lead(row: dict[str, Any]) -> Lead:
    metadata = row.get("metadata")
    if isinstance(metadata, (bytes, bytearray)):
        metadata = metadata.decode("utf-8", errors="ignore")
    if isinstance(metadata, str):
        try:
            metadata = json.loads(metadata)
        except json.JSONDecodeError:
            metadata = None
    return Lead(
        id=str(row["id"]),
        name=str(row.get("name") or ""),
        company=str(row.get("company") or ""),
        segment=row.get("segment"),
        email=str(row.get("email") or ""),
        whatsapp=str(row.get("whatsapp") or ""),
        objective=str(row.get("objective") or ""),
        monthly_volume=row.get("monthly_volume"),
        team_size=row.get("team_size"),
        current_tools=row.get("current_tools"),
        best_contact_time=row.get("best_contact_time"),
        source=str(row.get("source") or "landing"),
        status=str(row.get("status") or "new"),
        notes=row.get("notes"),
        assigned_to=str(row["assigned_to"]) if row.get("assigned_to") is not None else None,
        metadata=metadata if isinstance(metadata, dict) else None,
        created_at=row.get("created_at"),
        updated_at=row.get("updated_at"),
    )


class LeadService:
    def __init__(
        self,
        *,
        repository: LeadRepository,
        email_service: EmailService,
        settings: Settings,
    ) -> None:
        self.repository = repository
        self.email_service = email_service
        self.settings = settings

    async def create_lead(
        self,
        *,
        name: str,
        company: str,
        segment: str | None,
        email: str,
        whatsapp: str,
        objective: str,
        monthly_volume: str | None,
        team_size: str | None,
        current_tools: str | None,
        best_contact_time: str | None,
        source: str = "landing",
        metadata: dict[str, Any] | None = None,
    ) -> Lead:
        name = (name or "").strip()
        company = (company or "").strip()
        objective = (objective or "").strip()
        email_norm = _normalize_email(email)
        whatsapp_norm = _normalize_whatsapp(whatsapp or "")
        if not name or len(name) < 2:
            raise ValueError("Informe seu nome completo.")
        if not company:
            raise ValueError("Informe o nome da empresa.")
        if not _EMAIL_RE.match(email_norm):
            raise ValueError("E-mail invalido.")
        if not whatsapp_norm or len(whatsapp_norm) < 10:
            raise ValueError("WhatsApp invalido. Informe DDD + numero.")
        if not objective or len(objective) < 5:
            raise ValueError("Conte o que voce pretende automatizar no WhatsApp.")

        row = await self.repository.create(
            name=name,
            company=company,
            segment=(segment or None),
            email=email_norm,
            whatsapp=whatsapp_norm,
            objective=objective,
            monthly_volume=(monthly_volume or None),
            team_size=(team_size or None),
            current_tools=(current_tools or None),
            best_contact_time=(best_contact_time or None),
            source=(source or "landing").strip() or "landing",
            metadata=metadata,
        )
        if row is None:
            raise RuntimeError("Nao foi possivel salvar o lead.")
        lead = _row_to_lead(row)
        await self._notify_internal_team(lead)
        return lead

    async def list_leads(
        self,
        *,
        status: str | None = None,
        search: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> list[Lead]:
        rows = await self.repository.list_all(
            status=status,
            search=search,
            limit=limit,
            offset=offset,
        )
        return [_row_to_lead(row) for row in rows]

    async def get_status_summary(self) -> dict[str, int]:
        return await self.repository.count_by_status()

    async def update_lead(
        self,
        *,
        lead_id: int,
        status: str | None = None,
        notes: str | None = None,
        assigned_to: int | None = None,
    ) -> Lead:
        if status is not None and status not in _ALLOWED_STATUSES:
            raise ValueError(f"Status invalido: {status}")
        row = await self.repository.update_lead(
            lead_id=lead_id,
            status=status,
            notes=notes,
            assigned_to=assigned_to,
        )
        if row is None:
            raise ValueError("Lead nao encontrado")
        return _row_to_lead(row)

    async def _notify_internal_team(self, lead: Lead) -> None:
        recipient = (self.settings.BOOTSTRAP_COMPANY_ADMIN_EMAIL or "").strip()
        if not recipient or not self.email_service.configured:
            return
        subject = f"Novo lead Operada: {lead.company}"
        body = (
            f"Novo contato recebido em {datetime.utcnow().strftime('%d/%m/%Y %H:%M UTC')}.\n"
            f"Empresa: {lead.company}\n"
            f"Segmento: {lead.segment or '-'}\n"
            f"Contato: {lead.name}\n"
            f"E-mail: {lead.email}\n"
            f"WhatsApp: {lead.whatsapp}\n"
            f"Objetivo: {lead.objective}\n"
            f"Volume mensal: {lead.monthly_volume or '-'}\n"
            f"Ferramentas atuais: {lead.current_tools or '-'}\n"
            f"Melhor horario: {lead.best_contact_time or '-'}"
        )
        try:
            await self.email_service.send_billing_event_email(
                tenant_id=self.settings.BOOTSTRAP_COMPANY_TENANT_ID,
                tenant_name="Operada",
                to_email=recipient,
                subject=subject,
                body=body,
            )
        except Exception as exc:  # pragma: no cover - depends on SMTP
            logger.warning("Falha ao notificar time interno sobre lead %s: %s", lead.id, exc)
